[HPOS] Synchronize order deletions

Konamiman opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

1. Include order deletion in the HPOS synchronization process.
2. ~~Fix what was implemented in https://github.com/woocommerce/woocommerce/pull/36218 so that the upshifting only happens when the post type for the deleted order is hierarchical, as WordPress does when a post is deleted.~~ This part was moved to its own pull request: https://github.com/woocommerce/woocommerce/pull/38199
3. Add a data migration to delete stray order records that might be in the database (only the ones that can be deleted easily and safely). This process is performed in batches, 10k order rows are checked every in each step.

Also implemented but unrelated to the main purpose of the pull request:

1. `woocommerce_trash/delete_order` and `woocommerce_before_trash/delete_order` hooks are now fired in `Abstract_WC_Order_Data_Store_CPT::delete`, for consistency with `OrdersTableDataStore::delete`
2. `toggle_cot` method renamed to `toggle_cot_feature_and_usage` in `HPOSToggleTrait` and in `OrderHelper`.
3. Added the `order_exists` method in the `Abstract_WC_Order_Data_Store_CPT` and `OrdersTableDataStore` classes.
4. Added two new methods in `StringUtil`: `to_sql_list` and `class_name_without_namespace`.

#### Synchronization of order deletions

When an order is deleted and synchronization is enabled, the order is deleted from the backup tables too. This hasn't changed.

When an order is deleted and synchronization is disabled:

1. A record will be created in the `wp_wc_orders` table with `_deleted_from` as the key, and a fixed name representing the current authoritative table for orders (either `posts_table` or `orders_table`) as the value. This record marks the order as pending deletion from the backup table. The record will always be created in `wp_wc_orders`, regardless of which is the authoritative table.
4. Two new input values are defined for the `DataSynchronizer::get_ids_of_orders_pending_sync` method: `ID_TYPE_DELETED_FROM_ORDERS_TABLE` and `ID_TYPE_DELETED_FROM_POSTS_TABLE`.
5. The `_deleted_from` records will be included in the data returned by `DataSynchronizer::get_next_batch_to_process`. These records have the lowest priority (they will be returned only if there aren't enough added or modified orders to fill a batch of the requested size)
6. `DataSynchronizer::process_next_batch` will delete the orders that have a `_deleted_from` record with thea value that represents the current authoritative table.

An exception to this procedure is the case of an order deleted when the orders table is authoritative, while the corresponding records in the posts table are placeholders (because no sync happened between the order creation and its deletion). In this case, placeholder records are deleted immediately and no pending deletion records are created, even if sync is disabled.

#### Fix of the order upshifting

Now when an order is deleted from the orders table, upshifting of the child orders (setting the child orders parent id to the parent id of the order being deleted) will happen only when the post type of the deleted order is hierarchical, otherwise child orders will be deleted; [this is what WordPress does when a post is deleted](https://github.com/WordPress/WordPress/blob/master/wp-includes/post.php#L3419-L3426).

As of right now no order types are defined as hierarchical in WooCommerce, but this change in deletion behavior is covered by unit tests.

#### Data migration for stray order records

A data migration is added to delete the following records:

* ~~Records of type `shop_order_refund` with a parent order id of zero, from both the orders table and the posts table.~~ These won't be deleted from now, in case they were created by plugins.
* Placeholder records (post type `shop_order_placehold`) from the posts table, where there's no corresponding record (with the same id) in the orders table.

Deleting these records is safe. Deleting other possibly deleteable records (full orders still existing in the backup table) is more dangerous and will probably need a dedicated tool to be created in the future.

Closes #36580.

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

#### Testing overview

The testing basically consists of verifying that order deletion is properly handled (either deletion records are created in the meta table, or the corresponding records are deleted from the backup table for both the main order and any child orders) in all the possible scenarios. These scenarios are the combination of the following:

1. Orders table authoritative or posts table authoritative.
2. Automatic synchronization is enabled or disabled.
3. When automatic synchronization is disabled, the order is deleted before or after a manual synchronization is triggered.

Each test involves creating an order with a child order, the easiest way to do that is to create an order with a refund. From the admin area of the site:

1. Add a simple product (Products - Add New - set any name and price, then "Publish"); you need to do this only once.
2. Add an order (WooCommerce - Orders - Add order), add the product (Add item - Add product(s) - search the product by name and then "Add").
3. Click "Create" to actually create the order.
5. Add a refund: while in the order detaisl page, click "Refund", then in the quantity textbox for the item change the 0 to 1, and click the "Refund (amount) manually" button.
7. Click the "Update" button to update the order.

Deleting an order involves two steps: first, you trash the order ("Move to trash" link from the order edition page) and then you delete it ("Empty Trash" button from WooCommerce - Orders - Trash).

On the other hand, testing involves checking the state of the database. You can use WP CLI as follows (assuming `wp_` is the database prefix):

```
wp db query "select * from wp_posts where post_type like 'shop_order%'"
wp db query "select * from wp_postmeta"
wp db query "select * from wp_wc_orders"
wp db query "select * from wp_wc_orders_meta"
wp db query "select * from wp_wc_order_addresses"
wp db query "select * from wp_wc_order_operational_data"
```

If you start the testing with an empty site that should be enough, if not you'll need to add `WHERE` clauses to filter the results by the appropriate order and refund ids. "No order records exist" in the testing instructions means that the (possibly adjusted) above queries return no results.

Lastly, testing involves switching between the orders table and the posts table as the authoritative store for orders, and manually triggering synchronizations. You can do this via admin UI (WooCommerce - Settings - Advanced - Custom data stores) but again, WP CLI can help:

```
wp option set woocommerce_custom_orders_table_enabled yes
wp option set woocommerce_custom_orders_table_enabled no
wp wc cot sync
```

And with this we are ready:

#### Testing steps

1. Set the orders table as authoritative, with sync enabled.
2. Create an order with a refund, then delete it.
3. Verify that no order records exist.
4. Disable sync while keeping the orders table as authoritative.
6. Create an order with a refund, then delete it.
8. Verify that no order records exist (the corresponding post records had been created as placeholders, so they have been deleted even with sync disabled).
9. Create an order with a refund.
10. Manually trigger sync.
11. Delete the order.
12. Query the database and verify that no records exist in the orders tables, but the backup records do still exist in the posts table. Additionally, `_deleted_from` records have been created in `wp_wc_orders_meta` for both the order and the refund.
13. Manually trigger sync.
14. Verify that no order records exist.
15. Switch to posts table being authoritative, then repeat from step 2. The difference is that in step 6 no placeholder records will have been deleted (but still, no order records should exist: backup records hadn't been created to start with), and than in step 10 the backup records exist in the orders table this time.

Lastly, if you want to test the data migration you can do so by running the following in CLI:

```
wp eval 'require_once ABSPATH . "wp-content/plugins/woocommerce/includes/wc-update-functions.php"; do { $more = wc_update_800_delete_stray_order_records(); } while( $more );'
```

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
HPOS: Implement data synchronization for deleted orders.

This includes the following:

- Fix the "upshifting" that was implemented in 36218 so that it only
  applies when the post type of the order being deleted is hierarchical
  (this conforms to the WordPress behavior when deleting a post).

- Now when an order is deleted while sync is off a record will be
  created in wp_wc_orders_meta (regardless of which table is the
  authoritative one) with 'deleted_from' as the key and the
  authoritative table name (from where the order has been deleted)
  as the value...

- ...then DataSynchronizer will detect the presence of these
  'deleted_from' records and delete them from the backup table
  as part of the batch processing procedure.

- Exception to the above: when an order is deleted from the orders
  table and the corresponding records in the posts table are
  placeholders, these are deleted immediately too, even if sync is off.

- Add an "order_exists" method in both order data stores
  (Abstract_WC_Order_Data_Store_CPT and OrdersTableDataStore).

- Add a pair of get/set_verify_parent_id methods in WC_Abstract_Order
  (turning off the check in set_parent_id is necessary when deleting
  an order using the backup data store during synchronization).

- Placeholder records are now created with the appropriate parent id
  for child orders (e.g. refund), this is necessary in order to
  properly find which records need to be deleted.

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Add changelog file

Konamiman committed:
Add a data migration to delete stray order records from the database.

These are:

- Posts of type "shop_order_placeholder" with no matching order
  in the orders table.
- Records of type "shop_order_refund" with a zero parent order id
  in both tables.

Konamiman committed:
Add unit tests for the changes in DataSynchronizer

Konamiman committed:
And deletion sync related tests for OrdersTableDataStore.

Also:

- toggle_cot renamed to toggle_cot_feature_and_usage includes
  HPOSToggleTrais for clarity.
- DataSynchronizerTests and OrdersTableDataStoreTests now inherit
  from a common HposTestCase class.

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Rename more instances of "toggle_cot" to "toggle_cot_feature_and_usage"

Konamiman committed:
Add additional code sniffer fixes

vedanshujain commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2594,3 +2595,46 @@ function wc_update_750_disable_new_product_management_experience() {
+ *
+ * @return void
+ */
+function wc_update_770_delete_stray_order_records() {
```
I am not so sure about this migration, it's destructive and there is no check that the stray order records are coming because of HPOS. It could be expected behavior of some plugin or some merchant's bespoke setup.

Konamiman replied:
But the records that get deleted by this migration are unusable anyway: refunds with no parent order id (so there's no way to know which order do they belong and thus aren't used for anything), and placeholder records for which there's no matching order in `wp_wc_orders` (so they aren't "placeholding" anything).

vedanshujain replied:
Not sure about that, we can't say for sure if we created those refunds (instead of some plugin, or perhaps an import process). It's going towards very edge cases, but then there is also no harm in just keeping these records. wdyt?
Although, the placeholders records are fine, since we know we added them.

Konamiman replied:
I must have been doing something wrong before, but now I see that I can use `wc_get_order` for the child orders and I don't get the "Invalid parent ID" errors. So I have simplified the code accordingly (including removing the new `verify_parent_id` methods).

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/bb6645d4acd8e3b846c58b8a8ecc7ccf89cd4c1b

vedanshujain reviewed (commented):
So I have two major suggestions for flows that we currently have in this PR:

### Sync on delete

In this PR, we are including order deletion in the HPOS synchronization process. My suggestion is to add one more step before we get to this point, of confirming from users whether they do want to delete the orders. So basically, along with the line saying how many orders are pending sync, we can add one more line displaying that there are orders which are present in the backup table, but not in the authoritative table and confirm whether the user wants to delete them or not. 

The reason behind this suggestion is that currently we are transitioning everyone from posts based to HPOS, and many shops may enable/disable HPOS functionality by updating the option directly, (for example, if shops are facing issues then they may not want to wait for sync to finish). In such a case, we would be end up deleting all the non-migrated orders.

I understand this is bit of an edge case, and will cause issues with sync as well, but it's worthwhile giving another opportunity to opt out of deletions in the backup table. If we see no issues with this process, we can continue and include as part of normal sync.

### Having a `deleted_from` record.

Having the above intermediate step in place would mean that we can skip having a registry indicated by the `deleted_from` meta record. I am trying to think of a way where we would not need this registry at all, basically, this is just extra meta data make sense only in the context of sync but otherwise is not required in long term. Additionally, this would be something that folks working with WC would now need to keep in their mind, as they can;t delete entire records manually anymore.

WDYT?

Konamiman commented:
> we can add one more line displaying that there are orders which are present in the backup table, but not in the authoritative table and confirm whether the user wants to delete them or not.

We'll need to eventually add this too at some point, but that was intentionally left out of this PR.

So what we are doing here is providing a clean mechanism to prevent new stray undeleted record orders from appearing, and _after_ that is in place, we can think of a mechanism to allow users to do a manual cleanup. By doing it in that order, this manual cleanup will ideally be needed _only once_; which is a good thing because it will need some effort from the user to verify that the orders that are about to be deleted can actually be deleted.

Additionally, I think that database cleanup should happen closer to the point in which we'll make HPOS mandatory. The migration included in this PR is exceptional in that it deletes data that is clearly useless.

> this is just extra meta data make sense only in the context of sync but otherwise is not required in long term

In the long term we'll make HPOS mandatory, there won't be sync anymore and thus these records won't be created anymore; also as long as sync is working properly any existing deletion record will eventually disappear. This is actually a temporary solution.

> this would be something that folks working with WC would now need to keep in their mind, as they can;t delete entire records manually anymore

Indeed. But I think this is a small price to pay for the sake of database integrity (and as discussed, it's a temporary mechanism). Also this is just one more small thing to add to the list of "keep in mind when dealing with HPOS". 🙂

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Change data migration to target WC 7.8 instead of 7.7

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Remove deletion of zero parent refunds from the migration

Konamiman committed:
Change references to Woo 7.7 to 7.9

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Fix small space misalignment

Konamiman committed:
Change 'toggle_cot' class to 'toggle_cot_feature_and_usage'

Konamiman committed:
Add missing 'use' in HPOSToggleTrait

vedanshujain commented on plugins/woocommerce/includes/abstracts/abstract-wc-order.php:
```diff
@@ -103,6 +103,14 @@ abstract class WC_Abstract_Order extends WC_Abstract_Legacy_Order {
+	 * @since 7.9.0
+	 * @var bool
+	 */
+	private $verify_parent_id = true;
```
Seems like this is residual logic that was removed in #38199 , if so then let's remove it.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/5e1624be4662215f6fbf98d6bb6bccaff56f2a5d

vedanshujain commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2589,3 +2590,30 @@ function wc_update_770_remove_multichannel_marketing_feature_options() {
+
+	// phpcs:disable WordPress.DB.PreparedSQL
+
+	if ( ! $wpdb->get_var( $wpdb->prepare( 'SHOW TABLES LIKE %s', $wpdb->esc_like( $orders_table_name ) ) ) ) {
```
We would need to suppress errors here like so https://github.com/WordPress/wordpress-develop/blob/b98b5e59d72a68bc31f99b866fa26f7769d16e97/src/wp-admin/includes/upgrade.php#L2796-L2798 otherwise they will show in the $wpdb error log.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/ae1e340ae65da1d73e3727604e30c50e68351cff

vedanshujain commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2589,3 +2590,30 @@ function wc_update_770_remove_multichannel_marketing_feature_options() {
+
+	$wpdb->query(
+		$wpdb->prepare(
+			"DELETE FROM {$wpdb->posts} WHERE post_type = %s AND ID NOT IN (SELECT id FROM $orders_table_name)",
```
can we batch this, this seems like a pretty big query to do in an action scheduler time limit of 30 seconds. alternatively, can you please do some benchmarking around it so we know that this is not a problem even for large sites.

Konamiman replied:
How we would do batching here? I guess we would need to do the `SELECT` separately with a `LIMIT` and with a `WHERE id>x`, and store the highest obtained id to be used in the next batch as the `x`?

vedanshujain replied:
Yup, something like that would work. We can also use the `BatchProcessorInterface` and perhaps make improvements to it as we go. 

Unless we can test it out on some large site and prove that this isn't a problem that warrants batching efforts.

Konamiman replied:
I tried on a JN site with 100k orders created using smooth generator (orders authoritative, sync off, so 100k placeholders were created). I ran `time wp db query "DELETE FROM wp_posts WHERE post_type = 'shop_order_placehold' AND ID IN (SELECT id FROM wp_wc_orders)"` (changing the `ID NOT IN` to `ID IN` so that all the existing records would be affected).

The command took 10s to complete, so if the action limit is 30s it seems like yes, we'll need to batch.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/0759984b9fe560ec7018916834b82e900c13fd21 and fixed in https://github.com/woocommerce/woocommerce/pull/37050/commits/7a8891feedcd0fd004b586fbe750b566bfe98a8d

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
+			new \WC_Order_Data_Store_CPT() :
+			wc_get_container()->get( OrdersTableDataStore::class );
+
+		$order_ids_as_sql_list = '(' . implode( ',', $batch ) . ')';
```
let's map all values in `$batch` via `absint` function as well to prevent any possible SQLi. I don't know if we can trust `$batch` in all cases.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/c59b8cefad3ef827bde08fd79b73c35ae622ffec (in `process_batch`, which ultimately invokes `process_deleted_orders`).

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
+				continue;
+			}
+
+			if ( ! $data_store_for_deletion->order_exists( $order_id ) ) {
```
This check already happens for both data stores in the read method. We probably don't need another one separately, wdyt?

Konamiman replied:
If we leave the job to the data store `read` method we'll get a generic "invalid order id" error, while with this check we log a more explicit "Order doesn't exist in the backup table, thus it can't be deleted". So I prefer to leave this in place (also it makes the code more readable).

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
+				$order->set_id( $order_id );
+				$data_store_for_deletion->read( $order );
+
+				$data_store_for_deletion->delete( $order, array( 'force_delete' => true ) );
```
this will fire wc hooks again, perhaps we need some kind of suppress filter mechanism here?

Konamiman replied:
Maybe we could add a `suppress_firing_hooks` in the passed arguments array, and look for it in the `delete` method of both the CPT and the new data store class?

vedanshujain replied:
I think naming it `$suppress_filters` would keep it consistent with [existing WP_Query nomenclature](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/class-wp-query.php#L759). But otherwise yeah, that makes sense.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/406e2acba736357e1daeaa26631adc910e388ca0 for `OrdersTableDataStore`, and in https://github.com/woocommerce/woocommerce/pull/37050/commits/77af8532c62ea25b84eb93e55770150a7fea54e9 for `Abstract_WC_Order_Data_Store_CPT`.

Additionally, in https://github.com/woocommerce/woocommerce/pull/37050/commits/77af8532c62ea25b84eb93e55770150a7fea54e9 the  `Abstract_WC_Order_Data_Store_CPT` method has been modified to fire `woocommerce_before_delete/trash_order` actions, for symmetry with `OrdersTableDataStore`.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
+		if ( ! empty( $meta_ids_to_delete ) ) {
+			$order_id_rows_as_sql_list = '(' . implode( ',', $meta_ids_to_delete ) . ')';
+			// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+			$wpdb->query( "DELETE FROM {$wpdb->prefix}wc_orders_meta WHERE id IN {$order_id_rows_as_sql_list}" );
```
I know that only meta_key here would be a delete registry, but let's add a meta_key check anyways, incase someone has their own version of delete_registry implemented.

Konamiman replied:
On a second thought, this doesn't seem necessary. `$meta_ids_to_delete` is populated from the `$deletion_data` array that is obtained earlier in the same method using a SELECT query that is already filtering by `meta_key`. Or am I missing something?

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -389,17 +528,31 @@ public function get_total_pending_count(): int {
+		}
+
+		// Deleted orders pending sync have the lowest priority, and additionally,
+		// if such orders exist then we return just those, even if they don't fill a batch.
```
Is there a specific reason for this? Because otherwise, this will interfere with time remaining etc calculation in CLI (not that it's super accurate in the first place).

Konamiman replied:
The reason is the second part of the comment: "Returning 'missing orders that are to be deleted' and 'missing orders that are to be created' in the same batch is confusing and potentially dangerous". But isn't CLI using `get_current_orders_pending_sync_count` or something similar for displaying progress information?

vedanshujain replied:
> in the same batch is confusing and potentially dangerous

Looks like we have a check in the delete function already to look for `_delete_from` metadata, would you still say this could be dangerous? My reason for suggesting keeping the batch size the same is to be consistent with the current behavior, which can have orders to be inserted and to be updated as part of the same batch.

> get_current_orders_pending_sync_count

It does, but I think the percentage calculator assumes batch to have same size.

Konamiman replied:
Ok, `process_batch` is already telling apart deletion records from actual updates, so I guess we can change this part.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/0793d9da406625a7cd4769cdafc195e05e9b9cf0

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -451,9 +604,46 @@ public function get_description(): string {
+		}
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.SlowDBQuery
+		if ( $wpdb->get_var(
```
Seems like the first check will always be false when `$this->data_sync_is_enabled()` is true, and if it is false and posts are authoritative, we only need the first check and then insert directly, wdyt?

Konamiman replied:
By "the checks" you mean the `SELECT EXISTS` and the `AND NOT EXSITS`? The second one is there to prevent accidental creation of duplicate records.

vedanshujain replied:
Hmm, I was thinking along the lines of if we duplicate deleted_from records, would it really be a problem, but this check seems pretty lightweight, so it should be fine.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -451,9 +604,46 @@ public function get_description(): string {
+				$this->data_store::get_meta_table_name(),
+				array(
+					'order_id'   => $postid,
+					'meta_key'   => 'deleted_from',
```
should `deleted_from` be a constant defined somewhere? perhaps it should also start from `_` to mark it internal.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/e83eb705999548066ee8ad72c3d9abc53af13e3c

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1907,6 +1946,55 @@ public function delete( &$order, $args = array() ) {
+			$related_order_ids[] = $order_id;
+
+			// phpcs:disable WordPress.DB.SlowDBQuery
+			foreach ( $related_order_ids as $id ) {
```
picking it up in the select query when doing the sync using a where clause on `parent` would probably be faster then inserting for all children here, wdyt?

Konamiman replied:
Sorry, I'm not following you here. The deletion records need to be created precisely when sync is disabled, right? So what do you mean by "picking it up in the select query when doing the sync"?

vedanshujain replied:
So I mean that when [we select the orders that we want to delete](https://github.com/woocommerce/woocommerce/pull/37050/files/740e5e55a82708f3d04a6211f3ec44f675f3fa62#diff-1e070cc170d3b65cd6ba61b60cbd0697c8431fda674b4720926c8fbc1afecce7R365) here, we can add a clause also to pick children of those orders. That way we won't have to insert extra records during the deletion process (potentially making it slow). Since sync happens in the context of an action and not a request, a potentially slow query there is more acceptable.

Konamiman replied:
I think I get what you mean. So:

1. Don't create deletion records for child orders that need deletion.
2. Instead, at the actual deletion time, pick the child orders of those that have a deletion record and also delete them.

I understand the reasoning about the performance gain, but I don't like this approach because:

1. It makes things more difficult to understand. You reasoned that these deletion records are an extra moving part to keep track of; if additionally we change the logic to "child orders don't get a deletion record but need to be deleted too" it's even more intricate. (It also makes manual debugging more difficult: you can't get a glimpse of how many records are pending deletion by just counting the deletion records anymore)
2. It makes the code more complicated. `get_deleted_order_ids` has a `$limit` argument, how do we handle this when the limit includes only some of the childs of a given order? Also the code that counts pending deletions would need changes too, it's not just "count of deletion records" anymore.
3. I don't think the performance difference would be so big. How often do orders have childs and how many? These are usually refunds and an order will have at most a few of these, right?

vedanshujain replied:
Fair enough, one last request would be if we can insert them in bulk instead of looping over them and firing multiple queries?

Konamiman replied:
Ok, I think I can do that.

Konamiman replied:
Now I think I'm overcomplicating things. So two things can happen when deleting an order:

1. Upshifting is active for the order type. Then child orders will have their parent ids changed and thus no deletion records are needed for them.
2. Upshifting is *not* active for the order type. Then `upshift_or_delete_child_orders` will have deleted all the child orders, and this will have caused all the required deletion records for them to be created.

Thus, this `foreach` could be removed altogether, as there will be no remaining child orders at this point (all are either no longer children of the order, or they have been deleted and got their deletion record already). Does that make sense?

Konamiman replied:
Ok, change implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/49caf09db7b7b2c9c06f84846350e1905a51cfd4

vedanshujain reviewed (commented):
Great work! Added some comments, but overall looks good.

Konamiman committed:
Fix insertion of duplicated deletion records.

With orders table authoritative, sync disabled, an order with refunds
is deleted, a duplicate deletion record was being created for the refund.

That's because 'upshift_or_delete_child_orders' was deleting the child
orders, which caused a deletion record to be already created; and
after that, 'handle_order_deletion_with_sync_disabled' was again
creating a record for these child orders.

The fix is a bit convoluted because we can't know for sure if
'upshift_or_delete_child_orders' will delete the child orders or not.

Konamiman committed:
Deletion key renamed to '_deleted_from', and moved to constant.

Konamiman committed:
Use $wpdb->suppress_errors in the data migration

Konamiman committed:
Apply absint to the batch in DataSynchronizer::process_batch

Konamiman committed:
get_next_batch_to_process now returns changes of all types.

Previously it wouldn't mix "deleted" with "created" or "updated"
in the same batch.

Also dded the HposTestCase::set_order_as_updated method.

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Remove the "verify parent id" logic for orders

Konamiman committed:
Add a 'suppress_filters' arg to OrdersTableDataStore::delete

Also set it to true in DataSynchronizer::process_deleted_orders

Konamiman committed:
Add a 'suppress_filters' arg to Abstract_WC_Order_Data_Store_CPT::delete

Also this method now fires 'woocommerce_before_delete/trash_order'
hooks, for symmetry with OrdersTableDataStore::delete.

Konamiman committed:
Remove the explicit creation of deletion records for child orders.

See the comment added at the end of
OrdersTableDataStore::handle_order_deletion_with_sync_disabled

Konamiman committed:
Modify the stray order deletion migration to work in batches.

Also introduce some utility methods:

StringUtil::to_sql_list
StringUtil::class_name_without_namespace

Konamiman committed:
Fix a couple of @since tags

Konamiman committed:
Fix linting errors in StringUtilTest

Konamiman committed:
Migration fixed.

The first version was actually deleting all the existing placeholder
records.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -247,6 +251,16 @@ public function get_current_orders_pending_sync_count( $use_cache = false ): int
 		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
 		$pending_count = (int) $wpdb->get_var( $sql );
+
+		$deleted_from_table = $this->custom_orders_table_is_authoritative() ? $orders_table : $wpdb->posts;
```
I think we should use constants here for posts and orders table (instead of exact table name with wpdb prefix), it's possible that site's DB prefix may change (or for example, when a site is imported/exported)

Konamiman replied:
Isn't `posts` a property in the `$wpdb` object that returns the name of the posts table?

Or you mean `$orders_table`? Earlier in the method we're doing `$orders_table  = $this->data_store::get_orders_table_name();` which is the proper thing to do, right?

vedanshujain replied:
> Or you mean $orders_table? Earlier in the method we're doing $orders_table  = $this->data_store::get_orders_table_name(); which is the proper thing to do, right?

It is proper when we are fetching data, but I don't think that it's safe to store the exact table names, since the wpdb prefix can change in the future for a shop. If it changes, these records will be invalided.
However, if we store a constant instead, like `posts` or `orders` than it will be isolated from this change.

Konamiman replied:
Oh you mean the actual meta value that gets stored. Ok, that makes sense, I'll change it.

Konamiman replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/f589ecfd38c55ba7b44cee8ec43bab48171a6984 (and pull request description updated).

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -351,27 +395,123 @@ public function cleanup_synchronization_state() {
+				continue;
+			}
+
+			if ( ! $data_store_for_deletion->order_exists( $order_id ) ) {
```
[optional] perhaps we can do this in bulk, instead of looping through them one by one. currently this will execute 1 query per deleted record.

Konamiman replied:
This is a "should never occur" case (hence the warning) added for robustness, so I wouldn't spend time trying to improve it (unless we later discover that it actually happens often for some reason).

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -351,27 +395,123 @@ public function cleanup_synchronization_state() {
+			try {
+				$order = new \WC_Order();
+				$order->set_id( $order_id );
+				$data_store_for_deletion->read( $order );
```
[optional] we can use `read_mutliple` in case of order tables at least to make this faster.

Konamiman replied:
But we would then be leaking an implementation detail about `OrdersTableDataStore` that isn't relevant here and could cease to be true in the future (if the code is refactored), also it makes the code harder to understand ("why is `read_multiple` used here if there's just one order to read?").

Additionally, I don't think that compared to all the work that `read_multiple` does (especially the db accesses) saving an array creation and an extra function invocation would be really noticeable.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -451,9 +603,46 @@ public function get_description(): string {
-		if ( 'shop_order' === $post->post_type && $this->data_sync_is_enabled() ) {
+		global $wpdb;
+
+		if ( 'shop_order' !== $post->post_type && 'shop_order_refund' !== $post->post_type ) {
```
I think this should include all post types that is returned by `wc_get_order_types( 'cot-migration' )` and not just shop order and refunds.

Konamiman replied:
👍🏻 Good point, I'll change it.

Konamiman replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/37050/commits/834675d023dd5bf281351793c91b13cb30288872

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -451,9 +603,46 @@ public function get_description(): string {
+			return;
+		}
+
+		$features_controller = wc_get_container()->get( FeaturesController::class );
```
this should check if custom order tables exist, instead of the feature being enabled, we will enable the feature anyway for everyone with WC 8.0 (although we won't set HPOS as authoritative).

Konamiman replied:
But will the user still have the ability the turn off the feature?

If yes: we still need to check the feature anyway.
If not: do we need any check at all? Can't we just assume that the tables exist at this point?

Asking because I suspect that checking the feature is faster than checking if the table exists.

vedanshujain replied:
> I suspect that checking the feature is faster than checking if the table exists.

Makes sense, checking for tables will indeed be slow. I can link to this feature being enabled with tables being created, so we can keep this check as it is in that case.

vedanshujain reviewed (commented):
Awesome work! I left few minor comments, but overall this is testing well!

Konamiman committed:
Fix the post type check in DataSynchronizer::handle_deleted_post

Konamiman committed:
Use constant strings instead of table names for deletion records.

The meta values for deletion records will now be 'posts_table'
and 'orders_table', instead of the hardcoded table names. This is safer
because the database prefix could change (in case of a site migration,
for example).

Konamiman committed:
Merge branch 'trunk' into add/sync_of_deleted_orders

Konamiman committed:
Fix the wc_update_total_sales_counts function

It was relying on the 'before_delete_post' hook but it should rely
on 'woocommerce_before_delete_order' instead.

vedanshujain reviewed (approved):
Let's do it 🚢
